wandb/sdk/lib/sock_client.py

```diff
@@ -45,13 +50,32 @@ def shutdown(self, val: int) -> None:
     def set_socket(self, sock: socket.socket) -> None:
         self._sock = sock
 
+    def _sendall_with_timeout(self, data: bytes) -> None:
+        total_sent = 0
+        while total_sent < len(data):
+            start_time = time.monotonic()
+            try:
+                sent = self._sock.send(data[total_sent:])
+                # sent equal to 0 indicates a closed socket
+                if sent == 0:
+                    raise RuntimeError("socket connection broken")
```

Comment:
This is a case we will likely want to handle more gracefully in the future... connection broken will happen if the internal service process is killed or other non graceful shutdown paths.

Reply:
I think we will just return exit code 1 in the user process, I don't think we actually raise an exception?